### PR TITLE
feat: systemd LoadCredentialEncrypted support

### LINKED ISSUE #25: Phase 2: systemd LoadCredentialEncrypted support
## Summary

Add systemd `LoadCredentialEncrypted=` as a secret injection path, encrypting secrets at rest with TPM2 or host key binding. The `.env` file remains as a fallback — this is **additive, no breaking changes**.

**Predecessor:** Phase 1 — File permission hardening (Issue #22, PR #23) ✅

---

## Motivation

Currently, all 6 secret env vars live in plaintext in `.env`:
- `PROXMOX_TOKEN_ID`, `PROXMOX_TOKEN_SECRET`
- `OPNSENSE_API_KEY`, `OPNSENSE_API_SECRET`
- `SSH_USER`, `SSH_KEY_PATH`

Phase 1 hardened file permissions (0o600). Phase 2 eliminates plaintext at rest entirely by leveraging systemd's built-in credential encryption (`systemd-creds encrypt`), which binds secrets to the TPM2 chip and/or the host's machine key.

---

### How it works

1. `mcp-homelab install --use-credentials` encrypts each secret via `systemd-creds encrypt`
2. Encrypted credential files are stored in `/etc/credstore.encrypted/`
3. The systemd unit gets `LoadCredentialEncrypted=` directives for each secret
4. At service start, systemd decrypts credentials into `$CREDENTIALS_DIRECTORY/<name>`
5. `load_env()` in `config.py` checks `CREDENTIALS_DIRECTORY` first, reads files, falls back to `.env`

### Precedence chain (highest to lowest)

1. Environment variables already set (e.g., from shell)
2. `$CREDENTIALS_DIRECTORY/<var>` (systemd-injected)
3. `.env` file (existing behavior)

### Key function: `load_from_credentials_dir()`

New function in `config.py` that:
- Checks if `CREDENTIALS_DIRECTORY` env var is set
- Reads each expected secret from the directory
- Calls `os.environ.setdefault(name, value)` — does NOT override existing env vars
- Logs which secrets were loaded from credentials vs `.env`

---

## Scope Estimate (~230 LOC)

| Component | LOC | Notes |
|---|---|---|
| `config.py` — `load_from_credentials_dir()` | ~50 | New function, integrate into `load_env()` |
| `install.py` — `--use-credentials` flag | ~95 | Prompt for secrets, call `systemd-creds encrypt`, write files |
| `mcp-homelab.service` template | ~15 | Add `LoadCredentialEncrypted=` directives |
| Tests | ~70 | Mock CREDENTIALS_DIRECTORY, test precedence, test fallback |
| **Total** | **~230** | |

---

### Edge cases to handle

| Scenario | Behavior |
|---|---|
| systemd < 250 | `systemd-creds` doesn't exist. Detect, warn, skip — fall back to `.env` |
| Container / LXC (no TPM2) | Auto-detect, use `--with-key=host` instead of `tpm2` |
| Credential rotation | Must re-run `systemd-creds encrypt` for each changed secret |
| Hybrid mode | Some secrets in credentials, rest in `.env` — must work seamlessly |
| Non-systemd host | Feature is a no-op. `.env` path unchanged |

### Design decisions needed during implementation

1. **Encryption binding:** TPM2-first with host-key fallback, or always host-key? (Recommend: TPM2 with auto-fallback)
2. **Hybrid mode:** Allow partial credential sets? (Recommend: yes — `setdefault` handles this naturally)
3. **Container detection:** How to detect LXC/Docker to skip TPM2? (Recommend: check `/run/systemd/container` or `systemd-detect-virt`)
4. **UX flow:** Interactive prompts for secrets during install, or read from existing `.env`? (Recommend: read from `.env` if present, prompt for missing)
5. **Idempotency:** Re-running `--use-credentials` should re-encrypt, not fail
6. **Logging:** Log which source each secret came from at startup (credential dir vs .env vs env var)

---

## Acceptance Criteria

- [ ] `mcp-homelab install --use-credentials` encrypts secrets and configures systemd unit
- [ ] Server starts successfully reading from `$CREDENTIALS_DIRECTORY`
- [ ] Server falls back to `.env` when credentials dir is not available
- [ ] Hybrid mode works (some secrets in credentials, rest in `.env`)
- [ ] systemd < 250 detected and handled gracefully
- [ ] Container/LXC detected — uses host key instead of TPM2
- [ ] All existing tests still pass
- [ ] New tests cover: credential loading, precedence, fallback, edge cases
- [ ] No breaking changes to existing `.env`-only workflow

---

## Not in scope

- Secret rotation without restart (potential Phase 4)
- Remote keystore backends (Phase 3)
- Auto-rotation or lease management

### PR DESCRIPTION
**Pull request by asvarnon**
## Summary

Adds optional systemd `LoadCredentialEncrypted` support so API secrets can be encrypted at rest via `systemd-creds` and decrypted to tmpfs at service start, rather than sitting in a plaintext `.env` file.

Closes #25

### Core
- **config.py**: `_CREDENTIAL_KEYS` tuple (4 API secrets) and `load_from_credentials_dir()` — reads decrypted credentials from `CREDENTIALS_DIRECTORY` tmpfs, sets `os.environ`, per-variable source logging
- **server.py**: Startup chain updated: `load_env()` → `load_from_credentials_dir()` → `validate_env()`

### Installer
- **install.py**: `_encrypt_credentials()` — reads `.env`, creates `/etc/credstore.encrypted` (0o700, root-owned), runs `systemd-creds encrypt --with-key=auto` for each secret
- **install.py**: `_write_systemd_unit()` gains `credential_keys` parameter, injects `LoadCredentialEncrypted=` directives before `ExecStart=`
- **install.py**: `run_install()` — dynamic step counter (10 or 11 steps), conditional encryption step
- **cli.py**: `--use-credentials` flag on install command

### Security Hardening
- Idempotent `chown`/`chmod` on credstore directory after `mkdir`

### Tests (16 new, 522 total)
- 7 tests for `load_from_credentials_dir` (silent skip, missing dir, full load, override, fallback, whitespace, partial)
- 6 tests for `_encrypt_credentials` (all secrets, missing env, no secrets, subprocess failure, blank values, comments)
- 3 tests for `_write_systemd_unit` credentials (injection, no keys, container stripping)

### CODE REVIEW

File: mcp_homelab/core/config.py [hunk 1]
```diff
@@ -150,6 +150,62 @@ def load_env() -> None:
...
+
+    for key in _CREDENTIAL_KEYS:
+        credential_path = credentials_dir / key
+        if credential_path.is_file():
+            os.environ[key] = credential_path.read_text().strip()
+            logger.info("%s: loaded from credentials directory", key)
+            loaded_from_credentials += 1
+        elif os.environ.get(key):
+            logger.info("%s: loaded from environment/.env", key)
+            loaded_from_environment += 1
```
[the remark this answers is not in the document]
Comment (asvarnon) [reply mcp_homelab/core/config.py]: Accepted — added an if os.environ.get(key) guard at the top of the credential-loading loop so existing env vars are never overwritten. Precedence is now: shell env > credentials > .env. Also updated the docstring to document this.

File: mcp_homelab/setup/install.py [hunk 2]
```diff
@@ -164,39 +164,123 @@ def _update_server_config(config_path: Path, public_url: str) -> None:
...
+                text=True,
+                timeout=30,
+            )
+        except (OSError, subprocess.TimeoutExpired) as exc:
+            print(f"  \u2717 failed to encrypt {key}: {exc}", file=sys.stderr)
+            sys.exit(1)
+        if result.returncode != 0:
+            stderr = result.stderr.strip() if result.stderr else "(no output)"
+            print(f"  \u2717 failed to encrypt {key}: {stderr}", file=sys.stderr)
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (asvarnon) [reply mcp_homelab/setup/install.py]: Pushing back on this one. `--use-credentials` is an explicit opt-in to encrypted credential storage. If `systemd-creds` is missing or encryption fails, silently falling back to plaintext .env defeats the security intent — the user explicitly asked for encrypted credentials and should know immediately that it didn't work. The current behavior prints a clear error message explaining *what* failed and *why*, then exits. This is the right UX for a security-hardening flag: fail loud, don't silently downgrade.

File: mcp_homelab/setup/install.py [hunk 3]
```diff
@@ -164,39 +164,123 @@ def _update_server_config(config_path: Path, public_url: str) -> None:
...
+    if credential_keys:
+        credential_lines = "\n".join(
+            f"LoadCredentialEncrypted={key}" for key in credential_keys
+        )
+        # Insert credential directives before ExecStart=
+        rendered = rendered.replace(
+            "ExecStart=",
+            f"{credential_lines}\nExecStart=",
+            1,
+        )
```
[the remark this answers is not in the document]
Comment (asvarnon) [reply mcp_homelab/setup/install.py]: Accepted — good catch. When credential_keys is provided, _write_systemd_unit() now strips the EnvironmentFile= line from the rendered unit before injecting LoadCredentialEncrypted= directives. This prevents systemd from preloading .env secrets into process env, which would shadow the encrypted credentials via the precedence guard. Tests updated to include EnvironmentFile= in templates and assert removal in credential mode.
